meinberlin/apps/kiezradar/serializers.py

```diff
@@ -0,0 +1,111 @@
+        many=True,
+        allow_null=True,
+        required=False,
+    )
+    project_types = serializers.PrimaryKeyRelatedField(
+        queryset=ProjectType.objects.all(), many=True, allow_null=True, required=False
+    )
+    topics = serializers.PrimaryKeyRelatedField(
+        queryset=Topic.objects.all(), many=True, allow_null=True, required=False
+    )
+    status_display = serializers.CharField(source="get_status_display", read_only=True)
+
+    class Meta:
+        model = SearchProfile
+        fields = [
+            "id",
+            "user",
+            "name",
+            "description",
+            "disabled",
+            "status",
+            "status_display",
+            "query",
+            "query_text",
+            "organisations",
+            "districts",
+            "project_types",
+            "topics",
+        ]
+
+        read_only_fields = ["user"]
+
+    def create(self, validated_data):
+        # Pop many-to-many and one-to-many fields from validated_data
+        query_text = validated_data.pop("query_text", None)
+        organisations_data = validated_data.pop("organisations", None)
+        districts_data = validated_data.pop("districts", None)
+        project_types_data = validated_data.pop("project_types", None)
+        topics_data = validated_data.pop("topics", None)
+
+        if query_text:
+            query, _ = KiezradarQuery.objects.get_or_create(text=query_text)
+            validated_data["query"] = query
+
+        # need to create a profile before we can add m2m relations
+        search_profile = SearchProfile.objects.create(**validated_data)
+
+        if organisations_data:
+            search_profile.organisations.set(organisations_data)
+        if districts_data:
+            search_profile.districts.set(districts_data)
+        if project_types_data:
+            for project_type in project_types_data:
+                project_type, _ = ProjectType.objects.get_or_create(pk=project_type.id)
+                search_profile.project_types.add(project_type)
+        if topics_data:
+            search_profile.topics.set(topics_data)
+
+        return search_profile
+
+    def update(self, instance, validated_data):
+        """
+        Custom update to handle many-to-many relationships.
+        """
+        # Pop many-to-many fields from validated_data
+        query_text = validated_data.pop("query_text", None)
+        organisations_data = validated_data.pop("organisations", None)
+        districts_data = validated_data.pop("districts", None)
+        project_types_data = validated_data.pop("project_types", None)
+        topics_data = validated_data.pop("topics", None)
+
+        # Update one-to-many relationship
+        if query_text:
+            # remove existing query and assign or create another
+            query, _ = KiezradarQuery.objects.get_or_create(text=query_text)
+            validated_data["query"] = query
+
+        if organisations_data:
```

Comment:
same as for `create` above, the super call at the end already sets the m2m relations, so we don't need to pop and manually set them